File: java/BFS.java
// BFS: explores all neighbors at present depth before moving on to next depth level 

import java.util.*; 

public class BFS
  {
    public static void bfs(int[][] graph, int s, int V)
    {
      boolean[] visited = new boolean[V];
      LinkedList<Integer> queue = new LinkedList<>();

      visited[s] = true;
      queue.add(s);

      while(queue.size() != 0)
        {
          s = queue.poll();
          System.out.print( s + " ");

          for (int i = 0; i < V; i++)
            {
              if (graph[s][i] == 1 && !visited[i])
              {
                visited[i] = true;
                queue.add(i);
              }
            }
        }
    }
    public static void main(String[] args)
    {
      int graph[][] = {{0, 1, 1, 1, 0},
                        {1, 0, 1, 0, 0},
                        {1, 1, 0, 1, 1},
                        {1, 0, 1, 0, 1},
                        {0, 0, 1, 1, 0}};
      bfs(graph, 0, 5);
    }
  }

      

File: javascript/bfs.js
function bfs(graph, startNode)
{
  const visited = new Set();
  const queue = [startNode];

  while (queue.length > 0)
    {
      const currentVertex = queue.shift();
      console.log(currentVertex);
      visited.add(currentVertex);

      for(const neighbor of graph[currentVertex])
        {
          if(!visited.has(neighbor))
          {
            visited.add(neighbor);
            queue.push(neighbor);
          }
        }
    }
}
// Example usage:
const graph = {
  0: [1, 2],
  1: [2],
  2: [0, 3],
  3: [3]
};

bfs(graph, 0) // output = 0 1 2 3 
